openpype/modules/shotgrid/aop/patch.py

```diff
@@ -0,0 +1,52 @@
+from copy import copy
+
+
+from openpype.api import Logger
+from openpype.modules.shotgrid.lib import (
+    credentials,
+    settings,
+    server,
+)
+
+_LOG = Logger().get_logger("ShotgridModule.patch")
+
+
+def _patched_projects(
+    self, projection=None, only_active=True
+):
+    all_projects = list(self._prev_projects(projection, only_active))
+    if (
+        not credentials.get_local_login()
+        or not settings.filter_projects_by_login()
+    ):
+        return all_projects
+    try:
+        linked_names = _fetch_linked_project_names() or set()
+        return [x for x in all_projects if _upper(x["name"]) in linked_names]
+    except Exception as e:
+        print(e)
+        return all_projects
+
+
+def _upper(x):
+    return str(x).strip().upper()
+
+
+def _fetch_linked_project_names():
+    return {
+        _upper(x["project_name"])
+        for x in server.find_linked_projects(credentials.get_local_login())
+    }
+
+
+def patch_avalon_db():
```

Comment:
We need this feature to filter OP projects out against Shotgrid user <-> projects association. At the time of Avalon submodule, we had only two option available to tackle this business need. The first was to submit modification to the Avalon repository, it seemed to be clunky and time/effort consuming. The second option (chosen one) was to take an AOP-lite approach, with patching code in-place. The idea is to post filter `AvalonDB.projects()` against a prepared sg projects list. 

> If you need this behavior you should supply this change based on your "studio" specific addon but don't put it into module that should be merged into core.

I would like to avoid putting a thing into the core, but the core is so poorly thought and made of spaghetti that some stuff becomes hard to avoid.

Reply:
> but the core is so poorly thought and made of spaghetti that some stuff becomes hard to avoid

I understand, we did it on avalon side too, but this just can't be approved.

Still, you can move it out of the shotgrid module. You can create your studio module where this will happen, if you need it. We do it for our clients when they need their custom code.

> The idea is to post filter AvalonDB.projects() against a prepared sg projects list.

We can maybe change few things to make it possible but there are few questions. Is it user specific? If yes then it's impossible in v3. You want to hide it just on UI level or on backed level too? If just on UI then UI can be changed.

Reply:
> Is it user specific?
Yes and no, it's specific to a sg login that you leave in the settings
> You want to hide it just on UI level or on backed level too?
For now, we are talking about UI only

Reply:
> Yes and no, it's specific to a sg login that you leave in the settings

Ok, in that case we can't handle it without another hack in v3 "easily".

Please move the patch from the PR, since this modification of the class may affect other parts of code or other modules.